File: cpp/include/ReadInstances.cpp
#include "ReadInstances.hpp"
#include <fstream>
#include <iostream>
#include <sstream>

namespace ReadInstances {

std::vector<InstanceData> readInstances(const std::string &filePath) {
  std::ifstream file(filePath);
  std::vector<InstanceData> instances;

  if (!file.is_open()) {
    std::cerr << "[ERROR] Could not open file: " << filePath << std::endl;
    return instances;
  }

  std::string line;
  InstanceData current;
  bool readingValues = false;
  int valuesCount = 0;

  while (std::getline(file, line)) {
    // Skip empty lines
    if (line.empty())
      continue;

    // Check if it's an instance header
    if (line.find("# Instance") != std::string::npos) {
      // If we were reading an instance, save it before starting a new one
      if (readingValues && valuesCount == current.M) {
        instances.push_back(current);
      }

      // Reset for new instance
      current = InstanceData();
      readingValues = false;
      valuesCount = 0;
      continue;
    }

    // Read the parameters line (N K B optimalSum)
    if (!readingValues && line.find("#") == std::string::npos) {
      std::istringstream iss(line);
      if (iss >> current.M >> current.N >> current.B >> current.optimalSum) {
        if (current.M <= 0) {
          std::cerr << "[ERRO] Valor de M inválido: " << current.M << "\n";
          continue;
        }
        current.values.clear();
        readingValues = true;
      } else {
        std::cerr << "[ERRO] Linha mal formatada: " << line << "\n";
      }
      continue;
    }

    // Read the values line
    if (readingValues) {
      std::istringstream iss(line);
      partition::ValueType value;
      while (iss >> value) {
        current.values.push_back(value);
        valuesCount++;
      }

      // Check if we read all expected values
      if (valuesCount == current.N) {
        instances.push_back(current);
        readingValues = false;
      }
    }
  }

  // Don't forget to add the last instance if file ends without another header
  if (readingValues && valuesCount == current.M) {
    instances.push_back(current);
  } else if (readingValues) {
    std::cerr << "[ERRO] Instância incompleta ignorada.\n";
  }

  file.close();
  return instances;
}

// Função auxiliar para testar a leitura
void printInstances(const std::vector<InstanceData> &instances) {
  for (size_t i = 0; i < instances.size(); i++) {
    const auto &instance = instances[i];
    std::cout << "Instance " << (i + 1) << ":\n";
    std::cout << "  N: " << instance.M << ", M: " << instance.M
              << ", B: " << instance.B
              << ", optimalSum: " << instance.optimalSum << "\n";
    std::cout << "  Values: ";
    for (partition::ValueType value : instance.values) {
      std::cout << value << " ";
    }
    std::cout << "\n\n";
  }
}

} // namespace ReadInstances

File: python/include/read_instances.py
import os

class InstanceData:
    """Class to store the data of a single instance."""
    def __init__(self, M=0, N=0, B=0, optimal_sum=0, values=None):
        self.M = M                  # Number of elements
        self.N = N                  # Number of partitions
        self.B = B                  # Number of bits used to generate numbers
        self.optimal_sum = optimal_sum  # Known optimal partition sum
        self.values = values or []      # List of generated numbers

    def __repr__(self):
        """Short text representation (for debugging)."""
        return (f"InstanceData(M={self.M}, N={self.N}, B={self.B}, "
                f"optimal_sum={self.optimal_sum}, values={self.values[:5]}...)")



def read_instances(file_path="../instances/optimal_m_partition_instances.txt"):
    """
    Reads all instances from the given file.

    Expected file format:
        # Instance X
        M N B OPTIMAL_SUM
        num1 num2 num3 ...

    Args:
        file_path (str): Path to the instance file.

    Returns:
        list[InstanceData]: A list containing all parsed instances.
    """
    if not os.path.exists(file_path):
        print(f"[ERROR] File not found: {file_path}")
        return []

    instances = []
    current = None
    reading_values = False

    with open(file_path, "r") as file:
        for line in file:
            line = line.strip()
            if not line:
                continue  # Skip empty lines

            # Detect new instance header
            if line.startswith("# Instance"):
                # If we were reading an instance, store it before starting a new one
                if current and current.values:
                    instances.append(current)
                # Initialize a new instance
                current = InstanceData()
                reading_values = False
                continue

            # Read parameters line: M N B optimal_sum
            if not reading_values and not line.startswith("#"):
                parts = line.split()
                if len(parts) >= 4:
                    current.M = int(parts[0])
                    current.N = int(parts[1])
                    current.B = int(parts[2])
                    current.optimal_sum = int(parts[3])
                    reading_values = True
                continue

            # Read values (the array of numbers)
            if reading_values:
                values = list(map(int, line.split()))
                current.values.extend(values)

        # Add the last instance if file ends without a new header
        if current and current.values:
            instances.append(current)

    print(f"✅ {len(instances)} instances successfully read from {file_path}")
    return instances



def print_instances(instances):
    """
    Prints all instances to the console for verification.

    Args:
        instances (list[InstanceData]): List of instances to print.
    """
    for i, inst in enumerate(instances, start=1):
        print(f"Instance {i}:")
        print(f"  M: {inst.M}, N: {inst.N}, B: {inst.B}, optimal_sum: {inst.optimal_sum}")
        print(f"  Values: {' '.join(map(str, inst.values))}\n")

